blocks/settlement.go

```diff
@@ -0,0 +1,165 @@
+	a := b.ancestry.Load()
+	if a == nil {
+		b.log.Error(errBlockResettled.Error())
+		return fmt.Errorf("%w: block height %d", errBlockResettled, b.Height())
+	}
+	if !b.ancestry.CompareAndSwap(a, nil) { // almost certainly means concurrent calls to this method
+		b.log.Fatal("Block ancestry changed during settlement")
+		// We have to return something to keen the compiler happy, even though we
+		// expect the Fatal to be, well, fatal.
+		return errors.New("block ancestry changed during settlement")
+	}
+	close(b.settled)
+	return nil
+}
+
+// WaitUntilSettled blocks until either [Block.MarkSettled] is called or the
+// [context.Context] is cancelled.
+func (b *Block) WaitUntilSettled(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-b.settled:
+		return nil
+	}
+}
+
+func (b *Block) ancestor(ifSettledErrMsg string, get func(*ancestry) *Block) *Block {
+	a := b.ancestry.Load()
+	if a == nil {
+		b.log.Error(ifSettledErrMsg)
+		return nil
+	}
+	return get(a)
+}
+
+const (
+	getParentOfSettledErrMsg  = "Get parent of settled block"
+	getSettledOfSettledErrMsg = "Get last-settled of settled block"
+)
+
+// ParentBlock returns the block's parent unless [Block.MarkSettled] has been
+// called, in which case it returns nil.
+func (b *Block) ParentBlock() *Block {
+	return b.ancestor(getParentOfSettledErrMsg, func(a *ancestry) *Block {
+		return a.parent
+	})
+}
+
+// LastSettled returns the last-settled block at the time of b's acceptance,
+// unless [Block.MarkSettled] has been called, in which case it returns nil.
+// Note that this value might not be distinct between contiguous blocks.
+func (b *Block) LastSettled() *Block {
+	return b.ancestor(getSettledOfSettledErrMsg, func(a *ancestry) *Block {
+		return a.lastSettled
+	})
+}
+
+// Settles returns the executed blocks that b settles if it is accepted by
+// consensus. If `x` is the block height of the `b.ParentBlock().LastSettled()`
+// and `y` is the height of the `b.LastSettled()`, then Settles returns the
+// contiguous, half-open range (x,y] or an empty slice i.f.f. x==y. Every block
+// therefore returns a disjoint (and possibly empty) set of historical blocks.
+//
+// It is not valid to call Settles after a call to [Block.MarkSettled] on either
+// b or its parent.
+func (b *Block) Settles() []*Block {
+	return settling(b.ParentBlock().LastSettled(), b.LastSettled())
+}
+
+// IfChildSettles is similar to [Block.Settles] but with different definitions
+// of `x` and `y` (as described in [Block.Settles]). It is intended for use
+// during block building and defines `x` as the block height of
+// `b.LastSettled()` while `y` as the height of the argument passed to this
+// method.
+//
+// The argument is typically the return value of [LastToSettleAt], where that
+// function receives b as the parent. See the Example.
+func (b *Block) IfChildSettles(lastSettledOfChild *Block) []*Block {
+	return settling(b.LastSettled(), lastSettledOfChild)
+}
```

Comment:
I was pretty confused by this function. After staring at it for long enough, I realized that it is returning what blocks the child will settle if it is accepted.

Can we rename this to `ChildSettles`?

Also can we comment this it is not valid to call `(If)ChildSettles` if `Block.MarkSettled` has been called on this block?